mqtt/mqtt-bridge/src/persist/storage/ring_buffer/mod.rs

```diff
@@ -0,0 +1,416 @@
+mod block;
+pub mod error;
+mod mmap;
+
+use std::{
+    collections::VecDeque,
+    fs::{File, OpenOptions},
+    future::Future,
+    path::PathBuf,
+    pin::Pin,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc,
+    },
+    task::{Context, Poll, Waker},
+    time::Instant,
+};
+
+use crate::persist::storage::{
+    serialize::{binary_deserialize, binary_serialize, binary_serialize_size},
+    FlushOptions, FlushState, Storage, StorageResult,
+};
+use memmap::MmapMut;
+use parking_lot::Mutex;
+
+use std::io::Result as IOResult;
+
+use block::{calculate_hash, validate, BlockHeaderWithHash, Data};
+use error::RingBufferError;
+use mmap::{mmap_read, mmap_write};
+
+pub type RingBufferResult<T> = Result<T, error::RingBufferError>;
+
+#[derive(Default)]
+pub(crate) struct FilePointers {
+    write: AtomicUsize,
+    read: AtomicUsize,
+}
+
+/// Imagine there are three pointers:
+/// W - A write pointer
+/// R - A read pointer
+/// D - A delete pointer
+///
+/// With a file buffer similar to below:
+/// +---+---+---+---+---+---+---+---+---+
+/// | A | B | C | D | E | F | G | H | I |
+/// +---+---+---+---+---+---+---+---+---+
+///
+/// It is possible that W could wrap around
+/// and cause us to lose data due to variable
+/// sized data being written.
+/// If W + data size > R and should write precede read,
+/// the data could become corrupted.
+/// Either this is prevented by scanning the file for
+/// the next block and checking if the write will fit.
+/// Or, W + data must not pass both D and R.
+/// Not passing D helps with guaranteeing all data read was handled.
+/// Not passing R helps avoid corruption.
```

Comment:
This is the strategy I thought might work well in an async env where we don't want to lose any data we haven't handled yet (unless there is a bit flip or internal HW error that prevents getting the data). 

If there are any questions on this, let me know, I want to put rust docs through the code I am adding so it will be easy to re-read and test how to use.